### PR TITLE
Fixes blob memory allocation.

### PR DESCRIPTION
**Pull request by kelvinhammond**
Error reported here https://github.com/nodejs/node-addon-api/issues/1174#issuecomment-1423250863

This fixes the use after free memory issue with the blob object by creating a shared_ptr for the new blob's data that blob will then own.

### CODE REVIEW

**Comment by kelson42** [comment]
@kelvinhammond Which bug does it fox? There is no ticket linked to this PR.

**Comment by kelson42** [comment]
@kelvinhammond Great, this huge!!!

File: src/blob.h [hunk 1]
```diff
@@ -25,19 +23,30 @@ class Blob : public Napi::ObjectWrap<Blob> {
...
+
+      size_t size = 0;
+      std::shared_ptr<char> data = nullptr;
       if (info[0].IsArrayBuffer()) {  // handle ArrayBuffer
         auto buf = info[0].As<Napi::ArrayBuffer>();
-        auto refcontent_ =
-            std::string(static_cast<char *>(buf.Data()), buf.ByteLength());
+        size = buf.ByteLength();
+        data = std::shared_ptr<char>(new char[size]);
+        memset(data.get(), 0, size);
```
Comment (mgautierfr) [inline src/blob.h:33]: Is the `memset` really necassary as we will copy the data just after ?

**Review by mgautierfr — CHANGES_REQUESTED** [review]
It is still a pity that we have to copy the data but at least it should indeed fix the issue.
(And avoiding the copy is not easy)

I think we can remove the `memset` however.

File: src/blob.h (hunk 1 above)
Comment (kelvinhammond) [reply src/blob.h:33]: Not really, I just like to zero memory before use. I'll update it.

Comment (kelvinhammond) [reply src/blob.h]: Fixed in updated code

**Review by mgautierfr — APPROVED** [review]

**Pull request merged**